File: app/routers/crm_studio.py
"""CRM Studio router.

STU-011: Custom field CRUD endpoints under /ui/admin/studio/fields/{entity_type}.

Registered in app/main.py inside `if _S.crm_studio_enabled:` block.
"""
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request

from app.auth.deps import AuthenticatedUser
from app.auth.policy import enforce_cookie_csrf, require_admin_or_root, require_root
from app.auth.roles import Role, normalize_role
from app.models import (
    StudioFieldCreateIn,
    StudioFieldListOut,
    StudioFieldOut,
    StudioFieldUpdateIn,
)
import app.services.crm_studio_fields as fields_svc
from app.services.crm_studio_fields import (
    ALLOWED_ENTITY_TYPES,
    FieldAlreadyExistsError,
    FieldNotFoundError,
)

router = APIRouter(prefix="/ui/admin/studio", tags=["studio"])


def _item_to_out(item: dict) -> StudioFieldOut:
    return StudioFieldOut(
        entity_type=item.get("entity_type", ""),
        field_key=item.get("field_key", ""),
        label=item.get("label", ""),
        field_type=item.get("field_type", "text"),
        required=bool(item.get("required", False)),
        default_value=item.get("default_value"),
        picklist_name=item.get("picklist_name"),
        max_length=int(item.get("max_length", 1000)),
        min_value=float(item["min_value"]) if item.get("min_value") is not None else None,
        max_value=float(item["max_value"]) if item.get("max_value") is not None else None,
        sort_order=int(item.get("sort_order", 0)),
        is_active=bool(item.get("is_active", True)),
        created_by_sub=item.get("created_by_sub", ""),
        created_at=int(item.get("created_at", 0)),
        updated_by_sub=item.get("updated_by_sub", ""),
        updated_at=int(item.get("updated_at", 0)),
    )


@router.get("/fields/{entity_type}", response_model=StudioFieldListOut)
async def list_entity_fields(
    entity_type: str,
    include_inactive: bool = Query(False),
    cursor: Optional[str] = Query(None),
    limit: int = Query(100, ge=1, le=500),
    user: AuthenticatedUser = Depends(require_admin_or_root),
) -> StudioFieldListOut:
    if entity_type not in ALLOWED_ENTITY_TYPES:
        raise HTTPException(400, {"code": "invalid_entity_type", "entity_type": entity_type})
    # include_inactive requires ROOT
    if include_inactive and normalize_role(user.role) != Role.ROOT:
        raise HTTPException(403, {"code": "root_required_for_include_inactive"})

    items, next_cursor = fields_svc.list_fields(
        entity_type,
        include_inactive=include_inactive,
        limit=limit,
        cursor=cursor,
    )
    return StudioFieldListOut(
        fields=[_item_to_out(i) for i in items],
        next_cursor=next_cursor,
        total_count=len(items) if include_inactive and not cursor else None,
    )


@router.post("/fields/{entity_type}", response_model=StudioFieldOut, status_code=201)
async def create_entity_field(
    entity_type: str,
    request: Request,
    body: StudioFieldCreateIn,
    user: AuthenticatedUser = Depends(require_root),
) -> StudioFieldOut:
    enforce_cookie_csrf(request)
    if entity_type not in ALLOWED_ENTITY_TYPES:
        raise HTTPException(400, {"code": "invalid_entity_type", "entity_type": entity_type})
    try:
        item = fields_svc.create_field(
            user.sub,
            entity_type,
            body.field_key,
            body.label,
            body.field_type,
            required=body.required,
            default_value=body.default_value,
            picklist_name=body.picklist_name,
            max_length=body.max_length,
            min_value=body.min_value,
            max_value=body.max_value,
            sort_order=body.sort_order,
        )
    except FieldAlreadyExistsError:
        raise HTTPException(409, {"code": "field_already_exists", "field_key": body.field_key})
    except ValueError as e:
        raise HTTPException(400, {"code": str(e)})
    return _item_to_out(item)


@router.get("/fields/{entity_type}/{field_key}", response_model=StudioFieldOut)
async def get_entity_field(
    entity_type: str,
    field_key: str,
    user: AuthenticatedUser = Depends(require_admin_or_root),
) -> StudioFieldOut:
    item = fields_svc.get_field(entity_type, field_key)
    if not item:
        raise HTTPException(404, {"code": "field_not_found", "entity_type": entity_type, "field_key": field_key})
    return _item_to_out(item)


@router.patch("/fields/{entity_type}/{field_key}", response_model=StudioFieldOut)
async def update_entity_field(
    entity_type: str,
    field_key: str,
    request: Request,
    body: StudioFieldUpdateIn,
    user: AuthenticatedUser = Depends(require_root),
) -> StudioFieldOut:
    enforce_cookie_csrf(request)
    try:
        item = fields_svc.update_field(
            user.sub,
            entity_type,
            field_key,
            label=body.label,
            required=body.required,
            default_value=body.default_value,
            max_length=body.max_length,
            min_value=body.min_value,
            max_value=body.max_value,
            sort_order=body.sort_order,
            reactivate=body.reactivate,
        )
    except FieldNotFoundError:
        raise HTTPException(404, {"code": "field_not_found", "entity_type": entity_type, "field_key": field_key})
    return _item_to_out(item)


@router.delete("/fields/{entity_type}/{field_key}", status_code=204)
async def deactivate_entity_field(
    entity_type: str,
    field_key: str,
    request: Request,
    user: AuthenticatedUser = Depends(require_root),
) -> None:
    enforce_cookie_csrf(request)
    try:
        fields_svc.delete_field(user.sub, entity_type, field_key)
    except FieldNotFoundError:
        raise HTTPException(404, {"code": "field_not_found", "entity_type": entity_type, "field_key": field_key})

File: frontend/src/api/endpoints/crmStudio.ts
/**
 * SuiteCRM SECURITY/STUDIO (STU) cluster — API endpoint wrappers.
 *
 * Covers three LIVE backend routers (all admin/root-only):
 *   - ACL roles            /ui/admin/crm/acl-roles            (STU-002 / STU-003)
 *   - Security groups      /ui/admin/crm/security-groups      (STU-004)
 *   - Studio custom fields /ui/admin/studio/fields/...        (STU-011)
 *
 * Flags: crm_acl_enabled + crm_studio_enabled (default OFF → routers not
 * registered → 404 on every path). Callers handle the disabled state via
 * ApiError.status.
 *
 * TS interfaces are defined INLINE here (mirroring app/models.py) and exported.
 */
import { api } from "@/api/client";

// ─── ACL: permission matrix ──────────────────────────────────────────────────

/** Seven per-module boolean permission flags (CrmAclPermissionMatrix). */
export interface CrmAclPermissionMatrix {
  create?: boolean;
  read?: boolean;
  update?: boolean;
  delete?: boolean;
  export?: boolean;
  /** Reserved word in Python; serialized as "import" by the backend. */
  import?: boolean;
  mass_update?: boolean;
}

/** The seven canonical actions, in display order. */
export const ACL_ACTIONS: Array<{ key: keyof CrmAclPermissionMatrix; label: string }> = [
  { key: "create", label: "Create" },
  { key: "read", label: "Read" },
  { key: "update", label: "Update" },
  { key: "delete", label: "Delete" },
  { key: "export", label: "Export" },
  { key: "import", label: "Import" },
  { key: "mass_update", label: "Mass Update" },
];

/** Common CRM modules used by the ACL matrix (free-form on the backend). */
export const ACL_MODULES: string[] = [
  "contacts",
  "accounts",
  "leads",
  "opportunities",
  "cases",
  "projects",
  "tickets",
];

// ─── ACL: role models ────────────────────────────────────────────────────────

export interface CrmAclRole {
  role_id: string;
  name: string;
  description: string;
  permissions: Record<string, CrmAclPermissionMatrix>;
  created_at: number;
  created_by_sub: string;
  updated_at?: number | null;
  updated_by_sub?: string | null;
}

export interface CrmAclRoleCreateIn {
  name: string;
  description?: string;
  permissions?: Record<string, CrmAclPermissionMatrix>;
}

export interface CrmAclRoleUpdateIn {
  name?: string;
  description?: string;
  permissions?: Record<string, CrmAclPermissionMatrix>;
}

export interface CrmAclRoleListOut {
  roles: CrmAclRole[];
  next_cursor?: string | null;
}

export interface CrmAclAssignmentOut {
  role_id: string;
  user_sub: string;
  assigned_by_sub: string;
  assigned_at: number;
}

export interface CrmAclGroupAssignmentOut {
  role_id: string;
  group_key: string;
  assigned_by_sub: string;
  assigned_at: number;
}

export interface CrmAclRoleAssignmentsOut {
  role_id: string;
  /** Free-form user assignment dicts from the backend. */
  user_assignments: Array<Record<string, unknown>>;
  group_assignments: CrmAclGroupAssignmentOut[];
}

export interface CrmEffectivePermissionsOut {
  user_sub: string;
  permissions: Record<string, Record<string, boolean>>;
}

// ─── ACL: role endpoints ─────────────────────────────────────────────────────

const ACL_BASE = "/ui/admin/crm/acl-roles";

export const listAclRoles = (opts?: { cursor?: string; limit?: number }) => {
  const params: Record<string, string> = {};
  if (opts?.cursor) params["cursor"] = opts.cursor;
  if (opts?.limit) params["limit"] = String(opts.limit);
  return api.get<CrmAclRoleListOut>(`${ACL_BASE}/`, params);
};

export const getAclRole = (roleId: string) =>
  api.get<CrmAclRole>(`${ACL_BASE}/${encodeURIComponent(roleId)}`);

export const createAclRole = (body: CrmAclRoleCreateIn) =>
  api.post<CrmAclRole>(`${ACL_BASE}/`, body);

export const updateAclRole = (roleId: string, body: CrmAclRoleUpdateIn) =>
  api.patch<CrmAclRole>(`${ACL_BASE}/${encodeURIComponent(roleId)}`, body);

export const deleteAclRole = (roleId: string) =>
  api.del<void>(`${ACL_BASE}/${encodeURIComponent(roleId)}`);

export const getMyCrmPermissions = () =>
  api.get<CrmEffectivePermissionsOut>(`${ACL_BASE}/my-permissions`);

// ─── ACL: assignments ────────────────────────────────────────────────────────

export const listRoleAssignments = (roleId: string) =>
  api.get<CrmAclRoleAssignmentsOut>(`${ACL_BASE}/${encodeURIComponent(roleId)}/assignments`);

export const assignRoleToUser = (roleId: string, userSub: string) =>
  api.post<CrmAclAssignmentOut>(`${ACL_BASE}/${encodeURIComponent(roleId)}/assignments`, {
    user_sub: userSub,
  });

export const revokeRoleFromUser = (roleId: string, userSub: string) =>
  api.del<void>(
    `${ACL_BASE}/${encodeURIComponent(roleId)}/assignments/${encodeURIComponent(userSub)}`,
  );

export const assignRoleToGroup = (roleId: string, groupKey: string) =>
  api.post<CrmAclGroupAssignmentOut>(
    `${ACL_BASE}/${encodeURIComponent(roleId)}/group-assignments`,
    { group_key: groupKey },
  );

export const revokeRoleFromGroup = (roleId: string, groupKey: string) =>
  api.del<void>(
    `${ACL_BASE}/${encodeURIComponent(roleId)}/group-assignments/${encodeURIComponent(groupKey)}`,
  );

// ─── Security groups ─────────────────────────────────────────────────────────

export interface CrmSecurityGroup {
  group_id: string;
  name: string;
  description: string;
  owner_sub: string;
  created_at: number;
  is_global: boolean;
  member_count: number;
  record_count: number;
}

export interface CrmSecurityGroupMember {
  user_sub: string;
  added_by_sub: string;
  added_at: number;
  can_edit: boolean;
}

export interface CrmSecurityGroupRecord {
  entity_type: string;
  record_id: string;
  record_ref: string;
  assigned_by_sub: string;
  created_at: number;
}

export interface CrmSecurityGroupDetail {
  group: CrmSecurityGroup;
  members: CrmSecurityGroupMember[];
  records: CrmSecurityGroupRecord[];
}

export interface CrmSecurityGroupListOut {
  groups: CrmSecurityGroup[];
  next_cursor?: string | null;
}

export interface CrmSecurityGroupCreateIn {
  name: string;
  description?: string;
  is_global?: boolean;
}

const SG_BASE = "/ui/admin/crm/security-groups";

export const listSecurityGroups = (opts?: {
  cursor?: string;
  limit?: number;
  is_global?: boolean;
}) => {
  const params: Record<string, string> = {};
  if (opts?.cursor) params["cursor"] = opts.cursor;
  if (opts?.limit) params["limit"] = String(opts.limit);
  if (opts?.is_global !== undefined) params["is_global"] = String(opts.is_global);
  return api.get<CrmSecurityGroupListOut>(`${SG_BASE}/`, params);
};

export const getSecurityGroup = (groupId: string) =>
  api.get<CrmSecurityGroupDetail>(`${SG_BASE}/${encodeURIComponent(groupId)}`);

export const createSecurityGroup = (body: CrmSecurityGroupCreateIn) =>
  api.post<CrmSecurityGroup>(`${SG_BASE}/`, body);

export const deleteSecurityGroup = (groupId: string) =>
  api.del<void>(`${SG_BASE}/${encodeURIComponent(groupId)}`);

export const addSecurityGroupMember = (
  groupId: string,
  userSub: string,
  canEdit: boolean,
) =>
  api.post<void>(`${SG_BASE}/${encodeURIComponent(groupId)}/members`, {
    user_sub: userSub,
    can_edit: canEdit,
  });

export const removeSecurityGroupMember = (groupId: string, userSub: string) =>
  api.del<void>(
    `${SG_BASE}/${encodeURIComponent(groupId)}/members/${encodeURIComponent(userSub)}`,
  );

export const assignSecurityGroupRecord = (
  groupId: string,
  entityType: string,
  recordId: string,
) =>
  api.post<void>(`${SG_BASE}/${encodeURIComponent(groupId)}/records`, {
    entity_type: entityType,
    record_id: recordId,
  });

export const unassignSecurityGroupRecord = (
  groupId: string,
  entityType: string,
  recordId: string,
) =>
  api.del<void>(
    `${SG_BASE}/${encodeURIComponent(groupId)}/records/${encodeURIComponent(entityType)}/${encodeURIComponent(recordId)}`,
  );

// ─── Studio: custom fields ───────────────────────────────────────────────────

/** Entity types that accept custom fields (ALLOWED_ENTITY_TYPES). */
export const STUDIO_ENTITY_TYPES: string[] = [
  "contact",
  "ticket",
  "opportunity",
  "lead",
  "case",
  "project",
  "account",
];

export type StudioFieldType =
  | "text"
  | "integer"
  | "decimal"
  | "boolean"
  | "date"
  | "picklist"
  | "multi_picklist";

/** Field types that accept custom fields (ALLOWED_FIELD_TYPES). */
export const STUDIO_FIELD_TYPES: StudioFieldType[] = [
  "text",
  "integer",
  "decimal",
  "boolean",
  "date",
  "picklist",
  "multi_picklist",
];

export interface StudioField {
  entity_type: string;
  field_key: string;
  label: string;
  field_type: string;
  required: boolean;
  default_value: unknown;
  picklist_name?: string | null;
  max_length: number;
  min_value?: number | null;
  max_value?: number | null;
  sort_order: number;
  is_active: boolean;
  created_by_sub: string;
  created_at: number;
  updated_by_sub: string;
  updated_at: number;
}

export interface StudioFieldListOut {
  fields: StudioField[];
  next_cursor?: string | null;
  total_count?: number | null;
}

export interface StudioFieldCreateIn {
  field_key: string;
  label: string;
  field_type: StudioFieldType;
  required?: boolean;
  default_value?: unknown;
  picklist_name?: string | null;
  max_length?: number;
  min_value?: number | null;
  max_value?: number | null;
  sort_order?: number;
}

export interface StudioFieldUpdateIn {
  label?: string;
  required?: boolean;
  default_value?: unknown;
  max_length?: number;
  min_value?: number | null;
  max_value?: number | null;
  sort_order?: number;
  reactivate?: boolean;
}

const STUDIO_BASE = "/ui/admin/studio/fields";

export const listStudioFields = (
  entityType: string,
  opts?: { include_inactive?: boolean; cursor?: string; limit?: number },
) => {
  const params: Record<string, string> = {};
  if (opts?.include_inactive) params["include_inactive"] = "true";
  if (opts?.cursor) params["cursor"] = opts.cursor;
  if (opts?.limit) params["limit"] = String(opts.limit);
  return api.get<StudioFieldListOut>(`${STUDIO_BASE}/${encodeURIComponent(entityType)}`, params);
};

export const getStudioField = (entityType: string, fieldKey: string) =>
  api.get<StudioField>(
    `${STUDIO_BASE}/${encodeURIComponent(entityType)}/${encodeURIComponent(fieldKey)}`,
  );

export const createStudioField = (entityType: string, body: StudioFieldCreateIn) =>
  api.post<StudioField>(`${STUDIO_BASE}/${encodeURIComponent(entityType)}`, body);

export const updateStudioField = (
  entityType: string,
  fieldKey: string,
  body: StudioFieldUpdateIn,
) =>
  api.patch<StudioField>(
    `${STUDIO_BASE}/${encodeURIComponent(entityType)}/${encodeURIComponent(fieldKey)}`,
    body,
  );

export const deactivateStudioField = (entityType: string, fieldKey: string) =>
  api.del<void>(
    `${STUDIO_BASE}/${encodeURIComponent(entityType)}/${encodeURIComponent(fieldKey)}`,
  );
